### PR TITLE
param flag takes precedence over param-file

### PR DESCRIPTION
**Pull request by Himavanth**
## Description
Fixes https://github.com/adobe/aio-cli-plugin-runtime/issues/111
param flag takes precedence over param-file
annotation flag takes precedence over annotation-file

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Himavanth** [comment]
@shazron @purplecabbage The lone failing test here checks for a syntax that doesn't seem to be supported by `wsk` cli. Infact it throws an error. syntax in question is `wsk action invoke hello -p a b c d`. Should we still support this ?

**Comment by shazron** [comment]
> @shazron @purplecabbage The lone failing test here checks for a syntax that doesn't seem to be supported by `wsk` cli. Infact it throws an error. syntax in question is `wsk action invoke hello -p a b c d`. Should we still support this ?

we should not have any breaking changes and always follow what `wsk` does for now.

File: src/utils.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  const paramsActionObj = getKeyValueObjectFromMergedParameters(args, paramChar, paramString, paramFileChar, paramFileString
+    )
+  if (Object.keys(paramsActionObj).length > 0) {
+    return createKeyValueArrayFromObject(paramsActionObj)
+  } else {
+    return undefined
+  }
+}
+
+function getKeyValueObjectFromMergedParameters (args, paramChar, paramString, paramFileChar, paramFileString) {
```
Comment (purplecabbage) [inline src/utils.js:25]: Arg parsing should be handled by the oclif-inheriting commands.  There are a bunch of helper methods built in that make a lot of this trivial

**Comment by purplecabbage** [comment]
> > @shazron @purplecabbage The lone failing test here checks for a syntax that doesn't seem to be supported by `wsk` cli. Infact it throws an error. syntax in question is `wsk action invoke hello -p a b c d`. Should we still support this ?
> 
> we should not have any breaking changes and always follow what `wsk` does for now.

Because oclif just grabs all params and stuffs them in an array, we can be a bit more flexible and not consider this an error.
I think this ultimately depends on the other question of whether or not order matters ... assuming a params-file AND a param (-p & -P ) and both specify the same key with different values.  To me, a value specified on the command line overwrites anything specified in a file since it is more specific.  Just like ENV_VAR=5 ... should take precedence over any ENV_VAR defined in .env

**Comment by purplecabbage** [comment]
```
🔥🐝 aio app boom
flags =  {}
aio-cli-all/aio-cli-plugin-app
🔥🐝 aio app boom -p a v -p e k -p l r
flags =  { param: [ 'a', 'v', 'e', 'k', 'l', 'r' ] }
aio-cli-all/aio-cli-plugin-app
🔥🐝 aio app boom -p a v e k l r      
flags =  { param: [ 'a', 'v', 'e', 'k', 'l', 'r' ] }
aio-cli-all/aio-cli-plugin-app
🔥🐝 aio app boom -p a v -p e k l r
flags =  { param: [ 'a', 'v', 'e', 'k', 'l', 'r' ] }
aio-cli-all/aio-cli-plugin-app
🔥🐝 aio app boom -p a v -p e k -p r   
flags =  { param: [ 'a', 'v', 'e', 'k', 'r' ] }
```

**Comment by shazron** [comment]
> > > @shazron @purplecabbage The lone failing test here checks for a syntax that doesn't seem to be supported by `wsk` cli. Infact it throws an error. syntax in question is `wsk action invoke hello -p a b c d`. Should we still support this ?
> > 
> > 
> > we should not have any breaking changes and always follow what `wsk` does for now.
> 
> Because oclif just grabs all params and stuffs them in an array, we can be a bit more flexible and not consider this an error.
> I think this ultimately depends on the other question of whether or not order matters ... assuming a params-file AND a param (-p & -P ) and both specify the same key with different values. To me, a value specified on the command line overwrites anything specified in a file since it is more specific. Just like ENV_VAR=5 ... should take precedence over any ENV_VAR defined in .env

A bit more flexible may not align with our philosophy of 1:1 migration from `wsk`, if it doesn't break it, should be fine. In this case, it does, so we shouldn't.

**Comment by purplecabbage** [comment]
So it looks like there is agreement that the proposed solution of considering -p keys to override keys that come from the -P file.
This means we can adjust this to:
- parse -P file key:values into paramsObject
- parse all -p params and Object.assign them to the paramsObject ( overwriting in the process )

This further means that we can use oclif parsing of multiple flags and not have to reimplement it ourselves.
`flags =  { param: [ key, val, key, val, key, val ...] }`

**Pull request merged**